datadog-sidecar/src/service/sidecar_server.rs

```diff
@@ -432,215 +547,109 @@ impl SidecarInterface for SidecarServer {
+                                if client.buffered_composer_paths.insert(path.clone()) {
+                                    to_process.push(action);
+                                }
+                            }
+                            _ => {
+                                to_process.push(action);
+                            }
                         }
-                    };
-
-                    #[allow(clippy::expect_used)]
-                    manual_app_future
-                        .completer
-                        .expect("Completed expected Some ManualFuture for application instance, but found none")
-                        .complete(instance_option)
-                        .await;
-                }
-
-                if let Some(mut app) = manual_app_future.app_future.await {
-                    // Register metrics
-                    for metric in std::mem::take(&mut enqueued_data.metrics).into_iter() {
-                        app.register_metric(metric);
-                    }
-
-                    let mut actions: Vec<_> = std::mem::take(&mut enqueued_data.actions);
-
-                    // Send metric points
-                    for point in std::mem::take(&mut enqueued_data.points) {
-                        actions.push(app.to_telemetry_point(point));
                     }
 
-                    // drop on stop
-                    if actions.iter().any(|action| {
-                        matches!(action, TelemetryActions::Lifecycle(LifecycleAction::Stop))
+                    // Drop on Stop
+                    if to_process.iter().any(|action| {
+                        matches!(
+                            action,
+                            SidecarAction::Telemetry(TelemetryActions::Lifecycle(
+                                LifecycleAction::Stop
+                            ))
+                        )
                     }) {
-                        // Avoid self.get_runtime(), it could create a new one.
-                        if let Some(session) =
-                            self.sessions.lock_or_panic().get(&instance_id.session_id)
-                        {
-                            if let Some(runtime) =
-                                session.lock_runtimes().get(&instance_id.runtime_id)
-                            {
-                                runtime.lock_applications().remove(&queue_id);
+                        entry.remove();
+                    } else {
+                        let apps = rt_info.apps.clone();
+                        tokio::spawn(async move {
+                            let app_future =
+                                if let Some(fut) = apps.lock_or_panic().get(&(service, env)) {
+                                    fut.clone()
+                                } else {
+                                    return;
+                                };
+                            if let Some(mut app) = app_future.await {
```

Comment:
This is pending refactoring and the `telemetry_metrics` as well as the two methods from AppInstance is going to be moved TelemetryCachedClient, right?